File: langs/ruby/sequence.rb
#!/usr/bin/env ruby

def run_odometer(digits, base)
  odometer = Array.new(digits, 0)
  yield odometer
  high_value = base - 1
  digit = max_digit = digits - 1
  while digit > -1 do
    if odometer[digit] < high_value
      odometer[digit] += 1
      digit = max_digit
      yield odometer
    else
      odometer[digit] = 0
      digit -= 1
    end
  end
end

values = ['unknown', 'yes', 'no', 'maybe']

run_odometer(6, values.length) do |o|
  puts o.map {|i| "'#{values[i]}'"}.join(', ')
end

run_odometer(3, 10) do |o|
  puts o.inspect
end

# def generate_sequence(arr, num)
#   counter = Array.new(num, 0)
#   hi = arr.length - 1
#   pos = max_pos = counter.length - 1
#   yield counter
#   while pos > -1 do
#     if counter[pos] < hi
#       counter[pos] += 1
#       pos = max_pos
#       yield counter
#     else
#       counter[pos] = 0
#       pos -= 1
#     end
#   end
# end

# generate_sequence(['yes', 'no', 'maybe'], 4) do |s|
#   puts s.inspect
# end

File: langs/java/Sequence.java
public class Sequence {

    public static interface Listener {
        void tick(int[] arr);
    }

    public static void runOdometer(int digits, int base, Listener l) {
        int[] odometer = new int[digits];
        l.tick(odometer);
        int hiValue = base - 1;
        int maxDigit = digits - 1;
        int digit = maxDigit;
        while (digit > -1) {
            if (odometer[digit] < hiValue) {
                odometer[digit]++;
                digit = maxDigit;
                l.tick(odometer);
            } else {
                odometer[digit] = 0;
                digit--;
            }
        }
    }

    public static void main(String[] args) {
        final String[] values = new String[] {"unknown", "yes", "no", "maybe"};
        runOdometer(6, values.length, new Listener() {

            public void tick(int[] arr) {
                for (int i = 0; i < arr.length; i++) {
                    System.out.print("'" + values[arr[i]] + "'");
                    if (i < arr.length) System.out.print(", ");
                }
                System.out.println();
            }

        });
    }

}
